R/match_spec.R

```diff
@@ -67,106 +48,130 @@
 #' \code{\link{load_lib}()} loads the Open Specy reference library into an \R
 #' object of choice
 #'
-#' @importFrom rlang .data
-#' @importFrom stats approx cor
-#' @importFrom dplyr %>% inner_join mutate group_by ungroup summarize select arrange desc top_n
+#' @importFrom stats cor
+#' @importFrom data.table data.table setorder fifelse .SD
+#' @export
+cor_spec <- function(x, ...) {
+  UseMethod("cor_spec")
+}
+
+#' @rdname match_spec
+#'
+#' @export
+cor_spec.default <- function(x, ...) {
+  stop("object 'x' needs to be of class 'OpenSpecy'")
+}
+
+#' @rdname match_spec
+#'
+#' @export
+cor_spec.OpenSpecy <- function(x, library, na.rm = T, ...) {
+  if(sum(x$wavenumber %in% library$wavenumber) < 3){
+    stop("there are less than 3 matching wavenumbers in the objects you are trying to correlate, this won't work for correlation analysis. Consider first conforming the spectra to the same wavenumbers", call. = F)
+  }
+  cor(library$spectra[library$wavenumber %in% x$wavenumber,][,lapply(.SD, make_rel, na.rm = na.rm)][,lapply(.SD, mean_replace)],
+      x$spectra[x$wavenumber %in% library$wavenumber,][,lapply(.SD, make_rel, na.rm = na.rm)][,lapply(.SD, mean_replace)],
+      ...)
```

Comment:
Matching a spectrum with the complete library takes about 30 sec on my machine. Is this on purpose and my computer just too slow? Maybe we can speed this up somehow ..

Reply:
No that sounds right to me, it is just a slow algorithm, Ive been working on ways to speed it up using AI and K Mediods to reduce the size of the library. Also we could try approximate nearest neighbors instead of the exhaustive search.

Reply:
I did a little more testing: with about 15 secs each, applying `make_rel` and `mean_replace` to the complete library eats up most of the processing time. A simple solution to speed this up could be uploading a prenormalized library version to OSF and use this one by default. Are there any NAs in the library spectra that need replacement?

Reply:
Good find. There definitely are NAs, the issue is that we don't know which need to be replaced with the mean until after we do the cropping of the spectral range to match the unknown dataset so what we should replace the NAs with is different for each identification. We could remove the make_rel and probably still get similar performance because the correlation already centers and scales the data.

Reply:
Halving the processing time would already be great, I guess. Anyway, no need to touch it if the Shiny app performs okay .. what do you think?

Reply:
Just double checked, I think removing the make_rel may have weird side effects because correlation doesn't scale the data it just centers it. the new developments we are working on with model based Id and the library compression will end up having a really large effect on the speed so we can provide those as separate functions once we fine tune them. Some people just want traditional approach so I think we should have that as the default here.